doc/source/cluster/kubernetes/user-guides/pod-command.md

```diff
@@ -1,18 +1,69 @@
+Starting with KubeRay v1.1.0, if users add the annotation `ray.io/overwrite-container-cmd: "true"` to a RayCluster, KubeRay respects the container `command` and `args` as provided by the users, without including any generated command.
 
-## Timing 1: Before `ray start`
-Currently, for timing (1), we can set the container's `Command` and `Args` in RayCluster specification to reach the goal.
+```yaml
+apiVersion: ray.io/v1
+kind: RayCluster
+metadata:
+  annotations:
+    # If this annotation is set to "true", KubeRay will respect the container `command` and `args`.
+    ray.io/overwrite-container-cmd: "true"
+  ...
+spec:
+  headGroupSpec:
+    rayStartParams: {}
+    # Pod template
+    template:
+      spec:
+        containers:
+        - name: ray-head
+          image: rayproject/ray:2.8.0
+          # Because the annotation "ray.io/overwrite-container-cmd" is set to "true",
+          # KubeRay will overwrite the generated container command with `command` and
+          # `args` in the following. Hence, you need to specify the `ulimit` command
+          # by yourself to avoid Ray scalability issues.
+          command: ["/bin/bash", "-lc", "--"]
+          # Starting from v1.1.0, KubeRay injects the environment variable `KUBERAY_GEN_RAY_START_CMD`
+          # into the Ray container. This variable can be used to retrieve the generated Ray start command.
+          # Note that this environment variable does not include the `ulimit` command.
+          args: ["echo head; ulimit -n 65536; $KUBERAY_GEN_RAY_START_CMD"]
+          ...
+```
+
+The preceding example YAML is a part of [ray-cluster.overwrite-command.yaml](https://github.com/ray-project/kuberay/blob/master/ray-operator/config/samples/ray-cluster.overwrite-command.yaml).
+
+* `metadata.annotations.ray.io/overwrite-container-cmd: "true"`: This annotation tells KubeRay to respect the container `command` and `args` as provided by the users, without including any generated command.
+
+* `ulimit -n 65536`: This command is necessary to avoid Ray scalability issues caused by running out of file descriptors.
+If you don't set the annotation, KubeRay automatically injects the `ulimit` command into the container.
+
+* `$KUBERAY_GEN_RAY_START_CMD`: Starting from KubeRay v1.1.0, KubeRay injects the environment variable `KUBERAY_GEN_RAY_START_CMD` into the Ray container for both head and worker Pods to store the `ray start` command generated by KubeRay.
+Note that this environment variable doesn't include the `ulimit` command.
+
+The head Pod's `command`/`args` looks like the following:
+
+```yaml
+Command:
+  /bin/bash
+  -lc
+  --
+Args:
+  echo head; ulimit -n 65536; $KUBERAY_GEN_RAY_START_CMD
+```
+
+## Part 2: Execute commands before the generated `ray start` command
+
+If you only want to execute commands before the generated command, you don't need to set the annotation `ray.io/overwrite-container-cmd: "true"`.
```

Comment:
This part is a bit confusing. In the example above for Part 1, didn't we also only want to execute commands (`echo hello`) before the generated command? If that's the case, we should choose a different example for Part 1

Reply:
https://github.com/ray-project/kuberay/pull/1722